### PR TITLE
Bug 1992580: storageProfile should stay with the same value by check/uncheck the apply button

### PR DESCRIPTION
**Pull request by avivtur**
**Fixes**:
https://bugzilla.redhat.com/show_bug.cgi?id=1992580

**Analysis / Root cause**:
volume and access mode selector wasn't receiving any data as initial state

**Screen shots / Gifs for design review**:
before (taken from BZ) https://bugzilla.redhat.com/attachment.cgi?id=1813077 

after:

after

Signed-off-by: Aviv Turgeman <aturgema@redhat.com>

### CODE REVIEW

File: frontend/packages/console-app/src/components/volume-modes/volume-mode.tsx [hunk 1]
```diff
@@ -20,17 +20,24 @@ export const VolumeModeSelector: React.FC<VolumeModeSelectorProps> = (props) =>
     storageClass,
     availableVolumeMode,
     loaded,
+    spAccessMode,
+    spVolumeMode,
```
Comment (yaacov) [inline frontend/packages/console-app/src/components/volume-modes/volume-mode.tsx:24]:
```suggestion
    volumeMode,
```
can we just add initial value (e.g. volumeMode) and use ecisting access mode prop ?

File: frontend/packages/console-app/src/components/access-modes/access-mode.tsx [hunk 2]
```diff
@@ -33,6 +33,7 @@ export const AccessModeSelector: React.FC<AccessModeSelectorProps> = (props) =>
     availableAccessModes = [],
     description,
     ignoreReadOnly,
+    spAccessMode,
```
Comment (yaacov) [inline frontend/packages/console-app/src/components/access-modes/access-mode.tsx:36]:
```suggestion
    accessMode,
```
we just want to give this component an initial value, not a storage profile metadata, can we call the initial value `accessMode`

**Comment by avivtur** [comment]
@rawagner please review

File: frontend/packages/console-app/src/components/access-modes/access-mode.tsx (hunk 2 above)
Comment (avivtur) [reply frontend/packages/console-app/src/components/access-modes/access-mode.tsx:36]: accessMode is already taken by the state

File: frontend/packages/console-app/src/components/volume-modes/volume-mode.tsx (hunk 1 above)
Comment (avivtur) [reply frontend/packages/console-app/src/components/volume-modes/volume-mode.tsx:24]: also, volumeMode is already had it as the state
and I'll try to use the existing prop for accessMode

File: frontend/packages/console-app/src/components/access-modes/access-mode.tsx (hunk 2 above)
Comment (avivtur) [reply frontend/packages/console-app/src/components/access-modes/access-mode.tsx]: changed to `accMode`

File: frontend/packages/console-app/src/components/volume-modes/volume-mode.tsx (hunk 1 above)
Comment (avivtur) [reply frontend/packages/console-app/src/components/volume-modes/volume-mode.tsx]: changed to `volMode`
also using existing prop

File: frontend/packages/kubevirt-plugin/src/components/create-vm/forms/boot-source-form.tsx [hunk 3]
```diff
@@ -68,6 +68,9 @@ const AdvancedSection: React.FC<AdvancedSectionProps> = ({
 
   const [applySP, setApplySP] = React.useState<boolean>(true);
 
+  const isSCNameEmpty = !storageClassName || storageClassName === '';
```
Comment (pcbailey) [inline frontend/packages/kubevirt-plugin/src/components/create-vm/forms/boot-source-form.tsx:71]: See above. This shouldn't be necessary.

File: frontend/packages/kubevirt-plugin/src/components/create-vm/forms/boot-source-form.tsx [hunk 4]
```diff
@@ -142,14 +142,14 @@ const AdvancedSection: React.FC<AdvancedSectionProps> = ({
...
-            isDisabled={!isSPSettingProvided}
+            isDisabled={!isSPSettingProvided || isSCNameEmpty}
             label={t('kubevirt-plugin~Apply optimized StorageProfile settings')}
             data-test="apply-storage-provider"
           />
         </FormRow>
         {!spLoaded ? (
           <LoadingInline />
-        ) : isSPSettingProvided && applySP ? (
+        ) : isSPDefaults ? (
```
Comment (pcbailey) [inline frontend/packages/kubevirt-plugin/src/components/create-vm/forms/boot-source-form.tsx:152]: Since we're loading the same component for the `!spLoaded` and `!isSPdefaults` conditions, can we remove the nested ternary here and do like below?

```
{!spLoaded || !isSPDefaults ? (
  <LoadingInline />
) : (
    <Everything else>
)}
```
I think it would be even better if we can remove the negations and do this:

```
{spLoaded && isSPDefaults ? (
  <Everything else>
) : (
    <LoadingInline />
)}
```

File: frontend/packages/kubevirt-plugin/src/components/create-vm/forms/boot-source-form.tsx [hunk 5]
```diff
@@ -169,6 +169,9 @@ const AdvancedSection: React.FC<AdvancedSectionProps> = ({
                 provisioner={provisioner}
                 loaded
                 availableAccessModes={initialAccessModes}
+                accMode={
+                  isSPSettingProvided && !isSCNameEmpty ? spAccessMode?.getValue() : undefined
```
Comment (pcbailey) [inline frontend/packages/kubevirt-plugin/src/components/create-vm/forms/boot-source-form.tsx:173]: If I'm reading this correctly, `isSPDefaults` is true here, which means we already know that the conditional expression here is true. Am I missing something?

File: frontend/packages/kubevirt-plugin/src/components/create-vm/forms/boot-source-form.tsx [hunk 6]
```diff
@@ -180,9 +183,16 @@ const AdvancedSection: React.FC<AdvancedSectionProps> = ({
                   });
                 }}
                 provisioner={provisioner}
-                accessMode={state.accessMode?.value}
+                accessMode={
+                  isSPSettingProvided && applySP
```
Comment (pcbailey) [inline frontend/packages/kubevirt-plugin/src/components/create-vm/forms/boot-source-form.tsx:187]: Same issue as above, I think this condition should already be known to be true.

File: frontend/packages/kubevirt-plugin/src/components/create-vm/forms/boot-source-form.tsx [hunk 7]
```diff
@@ -180,9 +183,16 @@ const AdvancedSection: React.FC<AdvancedSectionProps> = ({
...
+                    ? spAccessMode?.getValue()
+                    : state.accessMode?.value
+                }
                 storageClass={storageClassName}
                 loaded
+                volMode={
+                  isSPSettingProvided && !isSCNameEmpty ? spVolumeMode?.getValue() : undefined
```
Comment (pcbailey) [inline frontend/packages/kubevirt-plugin/src/components/create-vm/forms/boot-source-form.tsx:194]: Same issue as above.

File: frontend/packages/kubevirt-plugin/src/components/modals/disk-modal/disk-modal.tsx [hunk 8]
```diff
@@ -136,8 +136,12 @@ export const DiskModal = withHandlePromise((props: DiskModalProps) => {
     storageClassName,
   );
 
+  const isSCNameEmpty = storageClassName === '' || !storageClassName;
```
Comment (pcbailey) [inline frontend/packages/kubevirt-plugin/src/components/modals/disk-modal/disk-modal.tsx:139]: See above. This shouldn't be necessary.

File: frontend/packages/kubevirt-plugin/src/components/modals/disk-modal/disk-modal.tsx [hunk 9]
```diff
@@ -717,15 +721,29 @@ export const DiskModal = withHandlePromise((props: DiskModalProps) => {
                           loaded
                           availableAccessModes={initialAccessModes}
                           description={accessModeHelp}
+                          accMode={
+                            isSPSettingProvided && !isSCNameEmpty
```
Comment (pcbailey) [inline frontend/packages/kubevirt-plugin/src/components/modals/disk-modal/disk-modal.tsx:725]: Same issue as above.

File: frontend/packages/console-app/src/components/access-modes/access-mode.tsx [hunk 10]
```diff
@@ -33,6 +33,7 @@ export const AccessModeSelector: React.FC<AccessModeSelectorProps> = (props) =>
     availableAccessModes = [],
     description,
     ignoreReadOnly,
+    accMode,
```
Comment (pcbailey) [inline frontend/packages/console-app/src/components/access-modes/access-mode.tsx:36]: nitpick: Could we name this something like `defaultAccessMode` or `initialAccessMode`?

File: frontend/packages/console-app/src/components/volume-modes/volume-mode.tsx [hunk 11]
```diff
@@ -20,14 +20,15 @@ export const VolumeModeSelector: React.FC<VolumeModeSelectorProps> = (props) =>
     storageClass,
     availableVolumeMode,
     loaded,
+    volMode,
```
Comment (pcbailey) [inline frontend/packages/console-app/src/components/volume-modes/volume-mode.tsx:23]: nitpick: Could we name this something similar to the above suggestions, like `defaultVolumeMode` or `initialVolumeMode`?

File: frontend/packages/console-app/src/components/access-modes/access-mode.tsx (hunk 10 above)
Comment (avivtur) [reply frontend/packages/console-app/src/components/access-modes/access-mode.tsx]: +1 changed to `initialAccessMode`

File: frontend/packages/console-app/src/components/volume-modes/volume-mode.tsx (hunk 11 above)
Comment (avivtur) [reply frontend/packages/console-app/src/components/volume-modes/volume-mode.tsx]: +1 changed to `initialVolumeMode`

File: frontend/packages/kubevirt-plugin/src/components/cdi-upload-provider/upload-pvc-form/upload-pvc-form.tsx [hunk 12]
```diff
@@ -153,13 +153,13 @@ export const UploadPVCForm: React.FC<UploadPVCFormProps> = ({
   const [spAccessMode, spVolumeMode, spLoaded, isSPSettingProvided] = useStorageProfileSettings(
     storageClassName || defaultSCName,
   );
+  const isSCNameEmpty = !storageClassName || storageClassName === '';
```
[the remark this answers is not in the document]
Comment (avivtur) [reply frontend/packages/kubevirt-plugin/src/components/cdi-upload-provider/upload-pvc-form/upload-pvc-form.tsx]: +1 removed `isSCNameEmpty`

File: frontend/packages/kubevirt-plugin/src/components/cdi-upload-provider/upload-pvc-form/upload-pvc-form.tsx [hunk 13]
```diff
@@ -511,7 +511,7 @@ export const UploadPVCForm: React.FC<UploadPVCFormProps> = ({
       </div>
       {!spLoaded ? (
         <LoadingInline />
-      ) : applySP && isSPSettingProvided ? (
+      ) : isSPDefaults ? (
```
[the remark this answers is not in the document]
Comment (avivtur) [reply frontend/packages/kubevirt-plugin/src/components/cdi-upload-provider/upload-pvc-form/upload-pvc-form.tsx:513]: see response below

File: frontend/packages/kubevirt-plugin/src/components/cdi-upload-provider/upload-pvc-form/upload-pvc-form.tsx [hunk 14]
```diff
@@ -526,14 +526,16 @@ export const UploadPVCForm: React.FC<UploadPVCFormProps> = ({
               provisioner={provisioner}
               loaded
               availableAccessModes={initialAccessModes}
+              accMode={isSPSettingProvided && !isSCNameEmpty ? spAccessMode?.getValue() : undefined}
```
[the remark this answers is not in the document]
Comment (avivtur) [reply frontend/packages/kubevirt-plugin/src/components/cdi-upload-provider/upload-pvc-form/upload-pvc-form.tsx]: see response below

File: frontend/packages/kubevirt-plugin/src/components/cdi-upload-provider/upload-pvc-form/upload-pvc-form.tsx [hunk 15]
```diff
@@ -526,14 +526,16 @@ export const UploadPVCForm: React.FC<UploadPVCFormProps> = ({
...
             />
           </div>
           <div className="form-group">
             <VolumeModeSelector
               onChange={(vMode) => setVolumeMode(vMode)}
               provisioner={provisioner}
-              accessMode={accessMode}
+              accessMode={applySP && isSPSettingProvided ? spAccessMode?.getValue() : accessMode}
```
[the remark this answers is not in the document]
Comment (avivtur) [reply frontend/packages/kubevirt-plugin/src/components/cdi-upload-provider/upload-pvc-form/upload-pvc-form.tsx]: changed to `accessMode || spAccessMode?.getValue()` to get initial value from SP when `accessMode` is `undefined`

File: frontend/packages/kubevirt-plugin/src/components/cdi-upload-provider/upload-pvc-form/upload-pvc-form.tsx [hunk 16]
```diff
@@ -526,14 +526,16 @@ export const UploadPVCForm: React.FC<UploadPVCFormProps> = ({
...
               storageClass={storageClassName}
+              volMode={isSPSettingProvided && !isSCNameEmpty ? spVolumeMode?.getValue() : undefined}
```
[the remark this answers is not in the document]
Comment (avivtur) [reply frontend/packages/kubevirt-plugin/src/components/cdi-upload-provider/upload-pvc-form/upload-pvc-form.tsx]: see response below

File: frontend/packages/kubevirt-plugin/src/components/create-vm/forms/boot-source-form.tsx (hunk 3 above)
Comment (avivtur) [reply frontend/packages/kubevirt-plugin/src/components/create-vm/forms/boot-source-form.tsx]: +1 removed `isSCNameEmpty`

File: frontend/packages/kubevirt-plugin/src/components/create-vm/forms/boot-source-form.tsx (hunk 4 above)
Comment (avivtur) [reply frontend/packages/kubevirt-plugin/src/components/create-vm/forms/boot-source-form.tsx:152]: there could be 2 different cases:
1. `spLoaded` is set to false which we are still waiting for data to be fetched from the StorageProfile Resource
2. `spLoaded` is set to true and this also has 2 different cases:
a. `isSPDeafults` is set to `false` which can happen either if the user unchecks the defaults OR there is no default 
values in StorageProfile OR when StorageClass is undefined, in this case, the access mode and volumes mode radio components should be visible to the user.
b. `isSPDeafults` is set to `true` in this case the default values from the StorageProfile should be visible

File: frontend/packages/kubevirt-plugin/src/components/create-vm/forms/boot-source-form.tsx (hunk 5 above)
Comment (avivtur) [reply frontend/packages/kubevirt-plugin/src/components/create-vm/forms/boot-source-form.tsx]: actually `isSPSettingProvided && !isSCNameEmpty` is not a trivial check to do, in line 159 there's a `:` sign that I assume you've missed which basically say that `isSPDefaults` is false so I'm checking if there are SP settings to give as the default values to the radio copmonents selection.

File: frontend/packages/kubevirt-plugin/src/components/create-vm/forms/boot-source-form.tsx (hunk 6 above)
Comment (avivtur) [reply frontend/packages/kubevirt-plugin/src/components/create-vm/forms/boot-source-form.tsx]: changed to `state.accessMode?.value || spAccessMode?.getValue()` to get initial value from SP when `state.accessMode?.value` is `undefined`

File: frontend/packages/kubevirt-plugin/src/components/create-vm/forms/boot-source-form.tsx (hunk 7 above)
Comment (avivtur) [reply frontend/packages/kubevirt-plugin/src/components/create-vm/forms/boot-source-form.tsx]: see response above

File: frontend/packages/kubevirt-plugin/src/components/modals/disk-modal/disk-modal.tsx (hunk 8 above)
Comment (avivtur) [reply frontend/packages/kubevirt-plugin/src/components/modals/disk-modal/disk-modal.tsx]: +1 removed `isSCNameEmpty`

File: frontend/packages/kubevirt-plugin/src/components/modals/disk-modal/disk-modal.tsx (hunk 9 above)
Comment (avivtur) [reply frontend/packages/kubevirt-plugin/src/components/modals/disk-modal/disk-modal.tsx]: see response above

File: frontend/packages/kubevirt-plugin/src/components/modals/disk-modal/disk-modal.tsx [hunk 17]
```diff
@@ -717,15 +721,29 @@ export const DiskModal = withHandlePromise((props: DiskModalProps) => {
...
+                          }
                         />
                       </StackItem>
                       <StackItem>
                         <VolumeModeSelector
                           onChange={(vMode) => setVolumeMode(VolumeMode.fromString(vMode))}
                           provisioner={storageProvisioner}
-                          accessMode={accessMode?.getValue()}
+                          accessMode={
+                            isSPSettingProvided && applySP
```
[the remark this answers is not in the document]
Comment (avivtur) [reply frontend/packages/kubevirt-plugin/src/components/modals/disk-modal/disk-modal.tsx]: changed to `accessMode?.getValue() || spAccessMode?.getValue()` to get initial value from SP when `accessMode?.getValue()` is `undefined`

File: frontend/packages/kubevirt-plugin/src/components/modals/disk-modal/disk-modal.tsx [hunk 18]
```diff
@@ -717,15 +721,29 @@ export const DiskModal = withHandlePromise((props: DiskModalProps) => {
...
+                              ? spAccessMode?.getValue()
+                              : accessMode?.getValue()
+                          }
                           storageClass={storageClassName}
                           loaded
+                          volMode={
+                            isSPSettingProvided && !isSCNameEmpty
```
[the remark this answers is not in the document]
Comment (avivtur) [reply frontend/packages/kubevirt-plugin/src/components/modals/disk-modal/disk-modal.tsx]: see response above

File: frontend/packages/kubevirt-plugin/src/components/create-vm/forms/boot-source-form.tsx (hunk 4 above)
Comment (pcbailey) [reply frontend/packages/kubevirt-plugin/src/components/create-vm/forms/boot-source-form.tsx:152]: Ah, I thought the false branch of the ternary that starts on line 198 was for the `isSPDefaults` condition. I didn't realize that it doesn't have a false branch. Have I mentioned how much I hate nested ternaries? lol. They're even worse when I can't select a bracket and have the matching bracket highlighted.

Would it work if you use `isSPDefaults && <Everything else>` instead of a ternary with only a true branch? I think that would make it more readable.

File: frontend/packages/kubevirt-plugin/src/components/create-vm/forms/boot-source-form.tsx [hunk 19]
```diff
@@ -180,9 +183,12 @@ const AdvancedSection: React.FC<AdvancedSectionProps> = ({
                   });
                 }}
                 provisioner={provisioner}
-                accessMode={state.accessMode?.value}
+                accessMode={state.accessMode?.value || spAccessMode?.getValue()} // Getting exitsting accessMode value otherwise
```
Comment (pcbailey) [inline frontend/packages/kubevirt-plugin/src/components/create-vm/forms/boot-source-form.tsx:186]: nitpick: exitsiting => existing

Sorry, I missed this one in the last round.

Comment (avivtur) [reply frontend/packages/kubevirt-plugin/src/components/create-vm/forms/boot-source-form.tsx:186]: no you didn't missed it, it's been added in last iteration but nice catch lol, I'll get it fixed

**Comment by yaacov** [comment]
@vojtechszocs @rawagner PTAL

**Comment by vojtechszocs** [comment]
LGTM @avivtur you can also update other files for consistency regarding the approach of pre-computing properties like `let spData` to avoid nested ternary operators which can be hard to follow when reading code.

**Comment by yaacov** [comment]
/hold

@avivtur hi, this need "yarn i18n", please remove the hold after running it

**Comment by avivtur** [comment]
@yaacov ran 'yarn i18n' , the `frontend` tests should be passing now
@vojtechszocs adjusted other files to follow consistency as you suggested, can you please take a look?
